src/FLSonar.cc

```diff
@@ -0,0 +1,689 @@
+// Copyright 2018 Brazilian Intitute of Robotics"
+
+#include <sstream>
+
+#include <ignition/math/Helpers.hh>
+#include <ignition/math/Pose3.hh>
+#include <ignition/math/Vector3.hh>
+
+#ifndef _WIN32
+#include <dirent.h>
+#else
+// Ensure that Winsock2.h is included before Windows.h, which can get
+// pulled in by anybody (e.g., Boost).
+#include <Winsock2.h>
+#include "gazebo/common/win_dirent.h"
+#endif
+
+#include "gazebo/rendering/ogre_gazebo.h"
+
+#include "gazebo/common/Assert.hh"
+#include "gazebo/common/Events.hh"
+#include "gazebo/common/Console.hh"
+#include "gazebo/common/Exception.hh"
+#include "gazebo/common/Mesh.hh"
+#include "gazebo/common/MeshManager.hh"
+#include "gazebo/common/Timer.hh"
+
+#include "gazebo/rendering/Visual.hh"
+#include "gazebo/rendering/Conversions.hh"
+#include "gazebo/rendering/Scene.hh"
+#include "foward_looking_gazebo_sonar/FLSonar.hh"
+
+using namespace gazebo;  // NOLINT(build/namespaces)
+using namespace rendering;  // NOLINT(build/namespaces)
+
+
+//////////////////////////////////////////////////
+FLSonar::FLSonar(const std::string &_namePrefix, ScenePtr _scene,
+                 const bool _autoRender)
+  : Camera(_namePrefix, _scene, false),
+    imageWidth(0),
+    imageHeight(0),
+    binCount(0),
+    beamCount(0)
+{
+}
+
+//////////////////////////////////////////////////
+FLSonar::~FLSonar()
+{
+  Ogre::TextureManager::getSingleton().remove(
+    this->camTexture->getName());
+}
+
+//////////////////////////////////////////////////
+void FLSonar::Load(sdf::ElementPtr _sdf)
+{
+  Camera::Load(_sdf);
+
+  GZ_ASSERT(_sdf->Get<double>("vfov"), "Vertical FOV is not set");
+
+  this->SetVertFOV(_sdf->Get<double>("vfov"));
+
+  Ogre::Radian fov_now(this->vfov);
+  this->camera->setFOVy(fov_now);
+  this->camera->setAspectRatio(1.0);
+  this->camera->setAutoAspectRatio(0);
+
+  {
+    sdf::ElementPtr iSdf = _sdf->GetElement("clip");
+    GZ_ASSERT(iSdf != nullptr, "clip is not set");
+    GZ_ASSERT(iSdf->Get<double>("far"), "far element of clip is not set");
+    this->SetFarClip(iSdf->Get<double>("far"));
+
+    GZ_ASSERT(iSdf->Get<double>("near"), "near element of clip is not set");
+    this->SetNearClip(iSdf->Get<double>("near"));
+  }
+
+  {
+    sdf::ElementPtr iSdf = _sdf->GetElement("image");
+    GZ_ASSERT(iSdf != nullptr, "Image is not set");
+    GZ_ASSERT(iSdf->Get<double>("width"), "Width element of image is not set");
+    this->SetImageWidth(iSdf->Get<double>("width"));
+
+    GZ_ASSERT(iSdf->Get<double>("height"), "Height element of image is not set");
+    this->SetImageHeight(iSdf->Get<double>("height"));
+  }
+
+  GZ_ASSERT(_sdf->Get<double>("bin_count"), "bin_count is not set");
+  this->SetBinCount(_sdf->Get<double>("bin_count"));
+
+  GZ_ASSERT(_sdf->Get<double>("beam_count"), "beam_count is not set");
+  this->SetBeamCount(_sdf->Get<double>("beam_count"));
```

Comment:
This stuff still doesn't look good. Why don't you factor this out in a helper static local function that does the assertion and returns the value of the element? Would look much better

Reply:
Changed